File: dnn-providers/hip-kernel-provider/rocKE/Python/rocke/instances/common/fmha_arch.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Arch-awareness helpers for the MFMA-tiled FMHA forward instances.

The MFMA-tiled attention body (:func:`rocke.helpers.mfma_attention.
mfma_attention_fwd_inner_body`) and the warp-distributed scalar bodies
in the ``fmha_*`` / ``sage_attention`` / ``sparse_attention`` instances
all run on the *narrow* MFMA atom catalog:

* f16 ``16x16x16`` (``MFMA_ATTN_BLOCK_M == MFMA_ATTN_BLOCK_K == 16``)
* bf16 ``16x16x16``

Both of those atoms exist on **gfx942 (CDNA3)** and **gfx950 (CDNA4)**,
so the canonical f16 / bf16 forward path is arch-portable. This helper
makes that fact explicit and machine-checkable: it sources the legal
atom set from :class:`rocke.core.arch.ArchTarget` (the same catalog
``instances/common/gemm_universal.is_valid_spec`` consults) and returns
a structured ``(ok, reason)`` instead of letting an arch-incompatible
config reach comgr (where a missing intrinsic HARD-CRASHES with
``LLVM ERROR: Cannot select intrinsic``).

The MFMA-tiled FMHA bodies never select the *wide* gfx950-only atoms
(``16x16x32`` / ``32x32x16``) on the f16 / bf16 path -- the helper pins
``BLOCK_K = 16`` -- so the validation below is a forward-looking guard
for the day a caller wires a wide-atom or fp8 attention variant through
these instances. Keeping the predicate arch-aware now means those
variants reject gfx942 cleanly instead of crashing the backend.
"""

from __future__ import annotations

from typing import Tuple


__all__ = [
    "FMHA_MFMA_ATTN_BLOCK",
    "validate_fmha_mfma_atom",
]


# The narrow attention MFMA tile both attention helpers pin (the K-dim
# of the QK / PV atom is 16; head_size is sliced into 16-wide atoms).
FMHA_MFMA_ATTN_BLOCK = 16


def validate_fmha_mfma_atom(dtype: str, arch: str = "gfx950") -> Tuple[bool, str]:
    """Return ``(ok, reason)`` for the MFMA-tiled FMHA atom on ``arch``.

    Mirrors :func:`rocke.instances.common.gemm_universal.is_valid_spec`'s
    catalog query: the f16 / bf16 ``16x16x16`` warp-tile atom the
    attention body uses must be in the target's MMA catalog. gfx942 and
    gfx950 both ship it, so the standard f16 / bf16 forward is accepted
    on both; an unknown arch (or a future dtype with no narrow atom on
    the target) is rejected with a structured reason.
    """
    from ...core.arch import ArchTarget

    try:
        target = ArchTarget.from_gfx(arch)
    except KeyError as e:
        return False, str(e)

    # Map the FMHA activation dtype string to the catalog dtype name.
    if dtype in ("f16", "fp16"):
        a_name = "f16"
    elif dtype == "bf16":
        a_name = "bf16"
    else:
        # validate_common_spec already restricts dtype to {f16,fp16,bf16};
        # this branch is defensive for callers that skip it.
        return False, f"unsupported FMHA dtype {dtype!r} for MFMA atom selection"

    blk = FMHA_MFMA_ATTN_BLOCK
    if not target.mma.has_shape(
        a_dtype=a_name, b_dtype=a_name, c_dtype="fp32", m=blk, n=blk, k=blk
    ):
        return False, (
            f"FMHA MFMA atom {a_name} {blk}x{blk}x{blk} not in {arch} catalog"
        )
    return True, "ok"

File: dnn-providers/hip-kernel-provider/rocke/platform/Cpp/instances/common/fmha_arch.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/*
 * C99 port of rocke/instances/common/fmha_arch.py:
 *   FMHA_MFMA_ATTN_BLOCK, validate_fmha_mfma_atom.
 * See rocke/helper_rocke.instances.common.fmha_arch.h for the symbol map.
 */
#include "rocke/helper_rocke.instances.common.fmha_arch.h"

#include <stdio.h> /* snprintf */
#include <string.h> /* strcmp   */

#include "rocke/arch_target.h" /* rocke_known_arches */
#include "rocke/helper_rocke.core.arch.h" /* rocke_archtarget_from_gfx, op_for_shape */

/* Copy a static reason into the caller buffer (NUL-terminated, truncating). */
static void rocke__set_reason(char* out, size_t out_cap, const char* reason)
{
    if(out == NULL || out_cap == 0)
    {
        return;
    }
    snprintf(out, out_cap, "%s", reason);
}

/* Reproduce str(KeyError(_build_target message)) for an unknown gfx target:
 *
 *   Python _build_target: raise KeyError(
 *     f"unknown gfx target {gfx!r}; known: {sorted(specs)}. "
 *     f"Add a row to {_DATA_FILE.name}.")
 *   validate_fmha_mfma_atom: return False, str(e)
 *
 * str(KeyError(msg)) == repr(msg); the message contains single quotes, so
 * Python's repr wraps it in DOUBLE quotes. sorted(specs) renders as a Python
 * list literal: ['gfx...', 'gfx...'] (single-quoted tokens, ", " separated).
 * known_arches() is already tuple(sorted(_load_specs())). */
static void rocke__set_unknown_arch_reason(char* out, size_t out_cap, const char* gfx)
{
    int count = 0;
    const char* const* arches;
    int i;
    size_t pos = 0;
    int wrote;

    if(out == NULL || out_cap == 0)
    {
        return;
    }

    arches = rocke_known_arches(&count);

    /* Leading double quote + the f-string up to "known: [" */
    wrote = snprintf(out + pos, out_cap - pos, "\"unknown gfx target '%s'; known: [", gfx);
    if(wrote < 0)
    {
        out[0] = '\0';
        return;
    }
    pos += (size_t)wrote;
    if(pos >= out_cap)
    {
        out[out_cap - 1] = '\0';
        return;
    }

    /* sorted(specs) list body: 'tok', 'tok', ... */
    for(i = 0; i < count; ++i)
    {
        wrote = snprintf(out + pos, out_cap - pos, "%s'%s'", (i == 0) ? "" : ", ", arches[i]);
        if(wrote < 0)
        {
            out[out_cap - 1] = '\0';
            return;
        }
        pos += (size_t)wrote;
        if(pos >= out_cap)
        {
            out[out_cap - 1] = '\0';
            return;
        }
    }

    /* Closing "]. Add a row to arch_specs.json." + trailing double quote. */
    snprintf(out + pos, out_cap - pos, "]. Add a row to arch_specs.json.\"");
}

bool rocke_validate_fmha_mfma_atom(const char* dtype, const char* arch, char* out, size_t out_cap)
{
    const char* a_name;
    const rocke_archtarget_t* target;
    int blk = ROCKE_FMHA_MFMA_ATTN_BLOCK;

    /* Python default: arch="gfx950". */
    if(arch == NULL)
    {
        arch = "gfx950";
    }

    /* try: target = ArchTarget.from_gfx(arch) except KeyError as e: return
     * False, str(e).  rocke_archtarget_from_gfx returns NULL for an unknown
     * target (the Python KeyError path). */
    target = rocke_archtarget_from_gfx(arch);
    if(target == NULL)
    {
        rocke__set_unknown_arch_reason(out, out_cap, arch);
        return false;
    }

    /* Map the FMHA activation dtype string to the catalog dtype name. */
    if(dtype != NULL && (strcmp(dtype, "f16") == 0 || strcmp(dtype, "fp16") == 0))
    {
        a_name = "f16";
    }
    else if(dtype != NULL && strcmp(dtype, "bf16") == 0)
    {
        a_name = "bf16";
    }
    else
    {
        /* return False, f"unsupported FMHA dtype {dtype!r} for MFMA atom
         * selection".  {dtype!r} renders single-quoted. */
        if(out != NULL && out_cap != 0)
        {
            snprintf(out,
                     out_cap,
                     "unsupported FMHA dtype '%s' for MFMA atom selection",
                     dtype != NULL ? dtype : "");
        }
        return false;
    }

    /* if not target.mma.has_shape(a_dtype=a, b_dtype=a, c_dtype="fp32",
     *                             m=blk, n=blk, k=blk): ...
     *
     * has_shape(m,n,k) is true iff a matching op exists; since enumerate()
     * filters on m,n and has_shape compares the full (m,n,k) shape, this is
     * exactly op_for_shape(...,k) != None. family defaults to "mma". */
    if(rocke_archtarget_op_for_shape(target, "mma", a_name, a_name, "fp32", blk, blk, blk) == NULL)
    {
        if(out != NULL && out_cap != 0)
        {
            snprintf(out,
                     out_cap,
                     "FMHA MFMA atom %s %dx%dx%d not in %s catalog",
                     a_name,
                     blk,
                     blk,
                     blk,
                     arch);
        }
        return false;
    }

    rocke__set_reason(out, out_cap, "ok");
    return true;
}
